File: apps/desktop-swift/Core/LocalUsage.swift
import Foundation

/// Local-AI usage ledger — portable logic (no SQLite, no AppKit).
///
/// Swift mirror of `apps/desktop-tauri/src-tauri/src/local_usage.rs`:
/// numbers only, never audio or transcript text. Units are honest STT units —
/// audio seconds in, words/chars out — because speech models have no LLM
/// tokens. Token slots stay reserved for a future local LLM.
///
/// Persistence lives in `Stores/LocalUsageStore.swift` (raw `sqlite3`, same
/// table as the Tauri app); this module owns the math and aggregation so both
/// stay unit-testable on any platform.
public enum LocalUsagePeriod: String, Sendable, CaseIterable {
    case today
    case sevenDays = "7d"
    case thirtyDays = "30d"
    case all
}

public struct LocalUsageRow: Sendable, Equatable {
    public var recordedAt: Date
    public var modelId: String
    public var engine: String
    public var audioSeconds: Double
    public var textChars: Int
    public var textWords: Int

    public init(
        recordedAt: Date,
        modelId: String,
        engine: String,
        audioSeconds: Double,
        textChars: Int,
        textWords: Int
    ) {
        self.recordedAt = recordedAt
        self.modelId = modelId
        self.engine = engine
        self.audioSeconds = audioSeconds
        self.textChars = textChars
        self.textWords = textWords
    }
}

public struct LocalModelUsage: Sendable, Equatable {
    public var modelId: String
    public var engine: String
    public var sessions: Int
    public var audioSeconds: Double
    public var textWords: Int
}

public struct LocalUsageSummary: Sendable, Equatable {
    public var sessions: Int
    public var audioSeconds: Double
    public var textChars: Int
    public var textWords: Int
    public var byModel: [LocalModelUsage]

    public static let empty = LocalUsageSummary(
        sessions: 0,
        audioSeconds: 0,
        textChars: 0,
        textWords: 0,
        byModel: []
    )
}

public enum LocalUsage {
    /// Unicode-whitespace word split — the same rule as Rust
    /// `split_whitespace` (`White_Space` property on both sides), so both
    /// clients count identically. Punctuation stays attached.
    public static func countWords(_ text: String) -> Int {
        text.split(whereSeparator: \.isWhitespace).count
    }

    /// Unicode scalar count (Rust `chars().count()`), not graphemes or bytes.
    public static func countChars(_ text: String) -> Int {
        text.unicodeScalars.count
    }

    /// UTC-based window start for a period (`nil` = unbounded). Missing maps
    /// to 30 days, matching the Rust default.
    public static func cutoff(
        for period: LocalUsagePeriod,
        now: Date = Date(),
        calendar: Calendar = utcCalendar()
    ) -> Date? {
        switch period {
        case .today:
            return calendar.startOfDay(for: now)
        case .sevenDays:
            return now.addingTimeInterval(-7 * 86_400)
        case .thirtyDays:
            return now.addingTimeInterval(-30 * 86_400)
        case .all:
            return nil
        }
    }

    public static func utcCalendar() -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")
            ?? TimeZone(secondsFromGMT: 0)
            ?? TimeZone.current
        return calendar
    }

    /// Totals + per-model breakdown (most sessions first) over `rows`.
    public static func summarize(
        rows: [LocalUsageRow],
        period: LocalUsagePeriod,
        now: Date = Date()
    ) -> LocalUsageSummary {
        let from = cutoff(for: period, now: now)
        let kept = from.map { start in rows.filter { $0.recordedAt >= start } } ?? rows
        var grouped: [String: LocalModelUsage] = [:]
        for row in kept {
            let key = "\(row.modelId)\n\(row.engine)"
            if var existing = grouped[key] {
                existing.sessions += 1
                existing.audioSeconds += row.audioSeconds
                existing.textWords += row.textWords
                grouped[key] = existing
            } else {
                grouped[key] = LocalModelUsage(
                    modelId: row.modelId,
                    engine: row.engine,
                    sessions: 1,
                    audioSeconds: row.audioSeconds,
                    textWords: row.textWords
                )
            }
        }
        let byModel = grouped.values.sorted { $0.sessions > $1.sessions }
        return LocalUsageSummary(
            sessions: kept.count,
            audioSeconds: kept.reduce(0) { $0 + $1.audioSeconds },
            textChars: kept.reduce(0) { $0 + $1.textChars },
            textWords: kept.reduce(0) { $0 + $1.textWords },
            byModel: byModel
        )
    }

    /// Human duration for audio totals: 45 → "45s", 150 → "2m 30s",
    /// 7500 → "2h 5m". Mirrors `formatAudioDuration` in the Tauri frontend.
    public static func formatAudioDuration(_ totalSeconds: Double) -> String {
        guard totalSeconds.isFinite, totalSeconds > 0 else { return "0s" }
        let total = Int(totalSeconds.rounded(.down))
        let hours = total / 3_600
        let minutes = (total % 3_600) / 60
        let seconds = total % 60
        if hours > 0 { return "\(hours)h \(minutes)m" }
        if minutes > 0 { return "\(minutes)m \(seconds)s" }
        return "\(seconds)s"
    }
}

File: apps/desktop-tauri/src/lib/local-usage.ts
import { invoke } from "@tauri-apps/api/core";
import { isTauri } from "./session/env.js";

export type LocalUsagePeriod = "today" | "7d" | "30d" | "all";

export interface LocalModelUsage {
  model_id: string;
  engine: string;
  sessions: number;
  audio_seconds: number;
  text_words: number;
}

export interface LocalUsageSummary {
  sessions: number;
  audio_seconds: number;
  text_chars: number;
  text_words: number;
  by_model: LocalModelUsage[];
}

export const EMPTY_LOCAL_USAGE: LocalUsageSummary = {
  sessions: 0,
  audio_seconds: 0,
  text_chars: 0,
  text_words: 0,
  by_model: [],
};

export const LOCAL_USAGE_PERIODS: LocalUsagePeriod[] = [
  "today",
  "7d",
  "30d",
  "all",
];

export function isLocalUsagePeriod(value: unknown): value is LocalUsagePeriod {
  return (
    typeof value === "string" &&
    (LOCAL_USAGE_PERIODS as string[]).includes(value)
  );
}

function isValidSummary(raw: unknown): raw is LocalUsageSummary {
  if (!raw || typeof raw !== "object") return false;
  const s = raw as Record<string, unknown>;
  return (
    typeof s.sessions === "number" &&
    typeof s.audio_seconds === "number" &&
    typeof s.text_chars === "number" &&
    typeof s.text_words === "number" &&
    Array.isArray(s.by_model) &&
    (s.by_model as unknown[]).every((m) => {
      if (!m || typeof m !== "object") return false;
      const row = m as Record<string, unknown>;
      return (
        typeof row.model_id === "string" &&
        typeof row.engine === "string" &&
        typeof row.sessions === "number" &&
        typeof row.audio_seconds === "number" &&
        typeof row.text_words === "number"
      );
    })
  );
}

export async function getLocalUsageSummary(
  period: LocalUsagePeriod,
): Promise<LocalUsageSummary> {
  // Browser preview and denied capabilities fall back to zeros — the ledger
  // is a local SQLite table that only exists in the desktop shell.
  if (!isTauri()) return EMPTY_LOCAL_USAGE;
  try {
    const summary = await invoke<unknown>("local_usage_summary", { period });
    return isValidSummary(summary) ? summary : EMPTY_LOCAL_USAGE;
  } catch {
    return EMPTY_LOCAL_USAGE;
  }
}

export async function clearLocalUsage(): Promise<number> {
  if (!isTauri()) return 0;
  try {
    const deleted = await invoke<number>("local_usage_clear");
    return typeof deleted === "number" && deleted >= 0 ? deleted : 0;
  } catch {
    return 0;
  }
}

/** Human duration for audio totals: 45 → "45s", 150 → "2m 30s", 7500 → "2h 5m". */
export function formatAudioDuration(totalSeconds: number): string {
  if (!Number.isFinite(totalSeconds) || totalSeconds <= 0) return "0s";
  const total = Math.floor(totalSeconds);
  const hours = Math.floor(total / 3600);
  const minutes = Math.floor((total % 3600) / 60);
  const seconds = total % 60;
  if (hours > 0) return `${hours}h ${minutes}m`;
  if (minutes > 0) return `${minutes}m ${seconds}s`;
  return `${seconds}s`;
}
